pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java

```diff
@@ -0,0 +1,901 @@
+          properties.get(FieldConfig.DERIVE_NUM_DOCS_PER_CHUNK_RAW_INDEX_KEY));
+    }
+    return false;
+  }
+
+  public static int rawIndexWriterVersion(String columnName, Map<String, Map<String, String>> columnProperties) {
+    if (columnProperties != null && columnProperties.get(columnName) != null) {
+      Map<String, String> properties = columnProperties.get(columnName);
+      String version = properties.get(FieldConfig.RAW_INDEX_WRITER_VERSION);
+      if (version == null) {
+        return BaseChunkSVForwardIndexWriter.DEFAULT_VERSION;
+      }
+      return Integer.parseInt(version);
+    }
+    return BaseChunkSVForwardIndexWriter.DEFAULT_VERSION;
+  }
+
+  /**
+   * Helper method that returns compression type to use based on segment creation spec and field type.
+   * <ul>
+   *   <li> Returns compression type from segment creation spec, if specified there.</li>
+   *   <li> Else, returns PASS_THROUGH for metrics, and SNAPPY for dimensions. This is because metrics are likely
+   *        to be spread in different chunks after applying predicates. Same could be true for dimensions, but in that
+   *        case, clients are expected to explicitly specify the appropriate compression type in the spec. </li>
+   * </ul>
+   * @param segmentCreationSpec Segment creation spec
+   * @param fieldSpec Field spec for the column
+   * @return Compression type to use
+   */
+  private ChunkCompressionType getColumnCompressionType(SegmentGeneratorConfig segmentCreationSpec,
+      FieldSpec fieldSpec) {
+    ChunkCompressionType compressionType = segmentCreationSpec.getRawIndexCompressionType().get(fieldSpec.getName());
+    if (compressionType == null) {
+      if (fieldSpec.getFieldType() == FieldType.METRIC) {
+        return ChunkCompressionType.PASS_THROUGH;
+      } else {
+        return ChunkCompressionType.SNAPPY;
+      }
+    } else {
+      return compressionType;
+    }
+  }
+
+  /**
+   * Returns true if dictionary should be created for a column, false otherwise.
+   * Currently there are two sources for this config:
+   * <ul>
+   *   <li> ColumnIndexCreationInfo (this is currently hard-coded to always return dictionary). </li>
+   *   <li> SegmentGeneratorConfig</li>
+   * </ul>
+   *
+   * This method gives preference to the SegmentGeneratorConfig first.
+   *
+   * @param info Column index creation info
+   * @param config Segment generation config
+   * @param spec Field spec for the column
+   * @return True if dictionary should be created for the column, false otherwise
+   */
+  private boolean createDictionaryForColumn(ColumnIndexCreationInfo info, SegmentGeneratorConfig config,
+      FieldSpec spec) {
+    String column = spec.getName();
+    if (config.getRawIndexCreationColumns().contains(column) || config.getRawIndexCompressionType()
+        .containsKey(column)) {
+      return false;
+    }
+    return info.isCreateDictionary();
+  }
+
+  @Override
+  public void indexRow(GenericRow row)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 34. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.